### ISSUE TITLE
SMPL mesh projection onto image

### ISSUE THREAD

**Issue by dqj5182** [wjrzm/MotionPRO#2]
Is it possible to project the SMPL mesh onto image?
I have tried projecting SMPL mesh onto image, but it seems like MotionPro dataset is just inherently not made to be well projected onto image.

To reproduce what I did, I inform you that I did my experiment for image `0814/zzb/2024-08-14-10-24-43/0/008227.jpg`.

```
# SMPL model
smpl_model = smplx.create('SMPL_NEUTRAL.pkl')

# Data info
orig_img_path = "0814/zzb/2024-08-14-10-24-43/0/008227.jpg"
date_name = orig_img_path.split('/')[-5]
seq_name = orig_img_path.split('/')[-4]
sub_seq_name = orig_img_path.split('/')[-3]
camera_id = orig_img_path.split('/')[-2]

# Load image
orig_img = cv2.imwrite(orig_img_path)

# Load camera
cam_param = {'focal': [607.46, 607.295], 'princpt': [642.213, 363.547]}
cam_extr = np.array([[ 0.51171153, -0.11485803,  0.85145044, -1.99804728],
       [-0.00827607, -0.99164185, -0.12875053,  1.31027417],
       [ 0.85911747,  0.05877139, -0.50838514,  2.2208739 ],
       [0., 0. ,0., 1.]])

# Body mesh
mesh_body_path = os.path.join(date_name, seq_name, sub_seq_name, 'smpl.npy')
mesh_body_data = np.load(mesh_body_path, allow_pickle=True).item()

betas = torch.tensor(mesh_body_data['betas'][:10], dtype=torch.float32)[None]
body_pose = torch.tensor(mesh_body_data['body_pose'], dtype=torch.float32)[image_id].reshape(-1, 23, 3)
global_orient = torch.tensor(mesh_body_data['global_orient'], dtype=torch.float32)[image_id][None, None]
transl = torch.tensor(mesh_body_data['transl'], dtype=torch.float32)[image_id][None]

mesh_smpl_out = smpl_model(betas=betas, # shape parameters
                            body_pose=body_pose, # pose parameters
                            global_orient=global_orient, # global orientation
                            transl=transl) # global translation

mesh_body_world = trimesh.Trimesh(mesh_smpl_out.vertices[0], smpl_model.faces)

# Transform to camera space
mesh_body_cam = mesh_body_world.copy()
mesh_body_cam.vertices = apply_homogeneous_transformation_np(mesh_body_cam.vertices, cam_extr)

# Project to image space
mesh_body_img = cam2pixel(mesh_body_cam.vertices, cam_param['focal'], cam_param['princpt'])

# Visualize 2D mesh vertices
cv2.imwrite('mesh_body_img.png', vis_keypoints(orig_img, mesh_body_img))
```

Note that I am using [vis_keypoints](https://github.com/mks0601/Hand4Whole_RELEASE/blob/9ba406cca06cf2f565b454b78991cc419e35023d/common/utils/vis.py#L57), [cam2pixel](https://github.com/mks0601/Hand4Whole_RELEASE/blob/9ba406cca06cf2f565b454b78991cc419e35023d/common/utils/transforms.py#L8), and [apply_homogeneous_transformation_np](https://github.com/dqj5182/HACO_RELEASE/blob/0f29b1fd6c6ddc9a336a8dc51635583d0bb146e0/lib/utils/transforms.py#L47).

However, this didn't work.
I observed `https://github.com/wjrzm/MotionPRO/blob/main/app/train_frappe.py` and you seemed to regard first and third axis as image plane in https://github.com/wjrzm/MotionPRO/blob/146a3774cca7807babe9e73c25b4aa6f497cf9d2/app/train_frappe.py#L37C86-L37C91. However, it seems that applying `mesh_body_cam.vertices = mesh_body_cam.vertices[:, [0, 2, 1]]` to the code above wouldn't work either.

May I ask if it is possible to project the annotated SMPL mesh onto image? or is it just impossible for this dataset?

**Comment by wjrzm** [comment]
First, the conclusion: Projecting SMPL meshes onto the image plane is feasible, but it inevitably introduces certain errors. These errors stem from: (1) limitations in optical motion capture precision, (2) discrepancies between the calculated origin of the camera coordinate system and the camera's true physical origin, and (3) minor inevitable translations and minimal rotational shifts in the camera setup due to significant time intervals (spanning two months) between capture sessions by different volunteers.

I am trying to debug your code and find details I have not clarified before:
### 1 The camera pose provided in Issue 1 #1 is the camera-to-world matrix; the camera pose needs to be inverted to obtain the world-to-camera matrix.
### 2 There's a -90-degree rotation between the SMPL coordinate system and the world coordinate system of the camera.

```
# SMPL model
smpl_model = smplx.create('SMPL_NEUTRAL.pkl')

# Data info
orig_img_path = "0814/zzb/2024-08-14-10-24-43/0/008227.jpg"
date_name = orig_img_path.split('/')[-5]
seq_name = orig_img_path.split('/')[-4]
sub_seq_name = orig_img_path.split('/')[-3]
camera_id = orig_img_path.split('/')[-2]

# Load image
orig_img = cv2.imwrite(orig_img_path)

# Load camera
cam_param = {'focal': [607.46, 607.295], 'princpt': [642.213, 363.547]}
cam_extr = np.array([[ 0.51171153, -0.11485803,  0.85145044, -1.99804728],
       [-0.00827607, -0.99164185, -0.12875053,  1.31027417],
       [ 0.85911747,  0.05877139, -0.50838514,  2.2208739 ],
       [0., 0. ,0., 1.]])

#1 The camera pose provided in Issue 1 is the camera-to-world matrix; 
#1 this matrix needs to be inverted to obtain the world-to-camera matrix.
cam_extr = np.linalg.inv(cam_extr)

# Body mesh
mesh_body_path = os.path.join(date_name, seq_name, sub_seq_name, 'smpl.npy')
mesh_body_data = np.load(mesh_body_path, allow_pickle=True).item()

betas = torch.tensor(mesh_body_data['betas'][:10], dtype=torch.float32)[None]
body_pose = torch.tensor(mesh_body_data['body_pose'], dtype=torch.float32)[image_id].reshape(-1, 23, 3)
global_orient = torch.tensor(mesh_body_data['global_orient'], dtype=torch.float32)[image_id][None, None]
transl = torch.tensor(mesh_body_data['transl'], dtype=torch.float32)[image_id][None]

mesh_smpl_out = smpl_model(betas=betas, # shape parameters
                            body_pose=body_pose, # pose parameters
                            global_orient=global_orient, # global orientation
                            transl=transl) # global translation

mesh_body_world = trimesh.Trimesh(mesh_smpl_out.vertices[0], smpl_model.faces)

# 2  Transform to world space 
# 2  There's a -90-degree rotation between the SMPL coordinate system and the world coordinate system of the camera.
angle_radians = np.radians(-90)
cos_theta = np.cos(angle_radians)
sin_theta = np.sin(angle_radians)
rotation_matrix = np.array([
    [1, 0,          0         ],
    [0, cos_theta, -sin_theta],
    [0, sin_theta,  cos_theta ]
])
mesh_body_world.vertices= mesh_body_world.vertices @ rotation_matrix.T

# Transform to camera space
mesh_body_cam = mesh_body_world.copy()
mesh_body_cam.vertices = apply_homogeneous_transformation_np(mesh_body_cam.vertices, cam_extr)

# Project to image space
mesh_body_img = cam2pixel(mesh_body_cam.vertices, cam_param['focal'], cam_param['princpt'])

# Visualize 2D mesh vertices
cv2.imwrite('mesh_body_img.png', vis_keypoints(orig_img, mesh_body_img))
```

The projected results might have some issues that aren't perfectly aligned with the image. I suggest slightly adjusting the translation parameters in the camera pose, as the problems mentioned initially often lead to subtle deviations in the origin of the camera coordinate system.

**Comment by dqj5182** [comment]
@wjrzm Thank you for your detailed answer!

I have tried your recommended solution and the SMPL mesh now does seem to be correct in terms of rotation. This is huge improvement from what I was experiencing.

However, it seems that `slightly adjusting the translation parameters in the camera pose` in camera space is more difficult than I expected.

## Not a single offset translation for each camera covers the entire dataset
To start with, I have manually annotated `offset translation` (for example, `np.array([-0.65, 0.05, 0.1])` for `cam0`), so that I can add it to SMPL vertices in camera space to make accurate 2D projection of 3D SMPL mesh.
These are the samples I used and corresponding `offset translation` that I used to add in camera space.
```
if camera_name == 'cam0': # aligned from 0814/zzb/2024-08-14-09-25-13/0/004319.jpg
    mesh_body_cam.vertices = mesh_body_cam.vertices + np.array([-0.65, 0.05, 0.1])
elif camera_name == 'cam1': # aligned from 0814/zzb/2024-08-14-11-30-54/1/009732.jpg
    mesh_body_cam.vertices = mesh_body_cam.vertices + np.array([-0.25, -0.15, 0.1])
elif camera_name == 'cam2': # aligned from 0814/zzb/2024-08-14-10-24-43/2/007289.jpg
    mesh_body_cam.vertices = mesh_body_cam.vertices + np.array([1.1, 0.2, -0.1])
elif camera_name == 'cam3': # aligned from 0814/zzb/2024-08-14-10-24-43/3/005800.jpg
    mesh_body_cam.vertices = mesh_body_cam.vertices + np.array([0.4, 0, 0])
else:
    raise NotImplementedError
```
This worked for those images. But, I found that these `offset translation` didn't work for other images. Especially, I found that certain `date_name` were not 2D aligned even after those `offset translation`.

So, I have checked based on `date_name` and found out that 
```
0729, 0905, 0906, 0909, 0910, 0912, 0913, 0914, 0919, 0920, 0923, 0925, 0926, 0927, 0929, 1008, 1009
```
did not work for the `offset translation` that I manually got while
```
0814, 0815, 0730, 0802, 0804, 0905, 0801, 0809, 0731, 0813, 0807, 0930, 0911, 0812, 0924
```
were working okay.

## Visualization
Before the manual `offset translation` of `np.array([0.4, 0, 0])` added to `0814/zzb/2024-08-14-10-24-43/3/005800.jpg` sample `mesh_body_cam.vertices`:

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/3f81102b-99fe-4af3-9158-a7dbd3c97604" />

After the manual `offset translation` of `np.array([0.4, 0, 0])` added to `0814/zzb/2024-08-14-10-24-43/3/005800.jpg` sample `mesh_body_cam.vertices`:

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/f2a76a78-8a44-479f-b4a3-c791b08041c2" />

Before the manual `offset translation` of `np.array([0.4, 0, 0])` from `0814/zzb/2024-08-14-10-24-43/3/005800.jpg` added to `0910/klf/2024-09-10-13-45-31/3/014167.jpg` sample `mesh_body_cam.vertices`:

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/51d280c9-361e-4714-91ae-21275b3e8a69" />

After the manual `offset translation` of `np.array([0.4, 0, 0])` from `0814/zzb/2024-08-14-10-24-43/3/005800.jpg` added to `0910/klf/2024-09-10-13-45-31/3/014167.jpg` sample `mesh_body_cam.vertices`:

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/d72cd213-0df1-4fc1-8bf2-1d1210f2df93" />

Note that `0910/klf/2024-09-10-13-45-31/3/014167.jpg` sample is new image sample that we experimented to add `offset translation` to check whether our `offset translation` of `np.array([0.4, 0, 0])` works for other sample.


## Better solution might be necessary for 2D alignment on all samples
I am okay with slight error in 2D projection as it is inevitable for most of the motion capture pipeline.
However, I wish to get (at leaset) globally aligned 2D projection which may be achievable by the `offset translation`.
But still, we could see that just a single `offset translation` for the entire dataset is not working. Also,  I have tried to manually annotate the `offset translation` for the problematic `date_name` of
```
0729, 0905, 0906, 0909, 0910, 0912, 0913, 0914, 0919, 0920, 0923, 0925, 0926, 0927, 0929, 1008, 1009
```
but not a single `offset translation` seemed to cover all of them.

May I ask if there could be more universal approach to get 2D aligned 3D mesh? or if not possible may I ask for release on `offset translation` data for each `date name` or if more is needed `offset translation` for each sample in the entire dataset?

**Comment by wjrzm** [comment]
@dqj5182 

Thanks for your testing. It appears there was some displacement change in the camera throughout the shooting cycle. The camera pose provided in issue1 #1  was captured on `1009`, and it might be feasible to directly apply that camera pose from issue1 #1 to dates like `1009` and its adjacent days. We will further examine the camera pose for each day. To expedite the resolution of your issue, we will prioritize processing the dates you manually annotated that didn't work: `0729, 0905, 0906, 0909, 0910, 0912, 0913, 0914, 0919, 0920, 0923, 0925, 0926, 0927, 0929, 1008, 1009`. Subsequently, we will gradually release data for all dates where we can compute the camera pose. 

Thank you for your attention and support—we hope this helps!

**Comment by dqj5182** [comment]
@wjrzm Thank you! I will wait for additional updates and also let you know if I have improvements on our end.

**Comment by wjrzm** [comment]
@dqj5182 We have compiled all available dates for calculating camera poses as much as possible. The link is here: https://drive.google.com/drive/folders/1I8mti28OKhBfgN7cUe5mE7p8HiBt-wMd?usp=sharing. For the missing dates, you can try using camera poses from nearby dates (with slight displacement changes if necessary). Hope this helps you.

**Comment by dqj5182** [comment]
Thank you for your help! It seems that most of the problems are solved, so I will close the issue.
